### ISSUE TITLE
Feature request: Inpainting masks or other way to mark the context for inpainting

### ISSUE THREAD

**Issue by Miraihi** [Acly/krita-ai-diffusion#38]
@Acly I really admire the creativite approaches you demonstrate in this project, but there's one thing that keep bugging me. You see, my major use for Stable Diffusion is cleaning text from manga comics, so the context is the king, and also I want to keep as much of the original image as possible intact. In [other](https://github.com/DrCyanide/cyanic-sd-krita)/earlier Krita SD plugins I'm always able to use a selection as a context, and then use the mask to mark the inpaiting area. But the way you plugin's inpainting works I'm very often starved of context.
"Padding area" has been a really nice addition that fixes the problem for many cases, but still sometime the working area may be really small conpared to the whole image, so even 25% of padding is insufficient.
I really like how flawlessly the plugin adapts the resolution for the task and how you can make the result look authentic and seamless, so I'm looking forward for the future improvements, whatever they may be.

**Comment by Acly** [comment]
It sounds like an interesting use case. If you could provide an example image that would help me understand better where current limitations are. Preferrably not a screenshot but maybe .kra with active selection so I can try it.

I never liked how inpaint required creating layer, drawing mask, and making selection, in other previous plugin. Seemed way too cumbersome. And I think generally automatic padding is good enough. So the question is how frequent is the situation where it isn't? Is it enough to allow more than 25% auto-padding? Is precise control over the context area required?

The context area is currently defined by the "selection bounds". A selection is a mask internally. The bounds are a box around non-transparent pixels. This actually gives you some control. I did the following:
* Make a selection of target area
* Krita automatically creates a "Selection Mask" layer, activate it
* Choose any small brush, and black color
* Mark (paint) the top-left and bottom-right corners of the context area you'd like
* Switch back to your paint layer, the mask becomes a selection again
* Generate

You can check by hovering over the Selection Mask layer in the layer docker, it shows a bigger area now. Meanwhile the black parts won't actually be inpainted because they have 0 transparency. (I thought you'd need at least 1, but apparently Krita tracks bounds separately after all). You will also see results include a lot of masked-out area which was part of generation.

I admit it feels a bit hacky, though not actually more clicks than a manual layer + selection would be. I haven't really come up with a more intuitive solution so far.

**Comment by Miraihi** [comment]
I'll surely try to create a quality illustration of what I mean later. A major point here is the way inpainting works under 100% strength. Essentially, if the area is too small it'll try to generate a whole image according to the prompt and put it inside the tiny selection area. The only way to remedy this right now is to mark several "dummy" areas at the other parts of the image to gather the context from there. (For example, if the text I want to remove is printed across the character's torso, you'll need to select a spot on their face and legs so the model have an idea that it works exactly with the torso).

**Comment by nakanakapoi** [comment]
I can't replicate the owner hacky workaround. Maybe I misunderstood the steps. When I select the selection mask, then the area outside the selection turn red, then I 'mark' my desired context area? This did nothing for me.

Context area is very important for human anatomy. When I want to correct the hand, I select the whole figure, then erase all the generated area except the hand. Many time I just 'image to image' the whole picture, then erase everything else. If I just to select the hand, I must significantly lower the strength and modified the prompt. However, the hand will be over-detailed and not match the rest of the picture.

I get it that previous approach by similar plugin (auto-sd-paint-ext) is cumbersome. Yours is more user-friendly like Adobe. I also see that you have some trick under the hood to make thing seamless, like ipadapter and stuff. But for power user like me, the ability to manually select the context area is very useful. If the other plugin has not stalled, I will probably stick to it for this one feature.

However, this is no big deal because a lot of workarounds exist. I did not expect you to change trajectory because this seem like a complete overhaul. I'm just happy that this plugin exists.

**Comment by nakanakapoi** [comment]
I finally able to do your hacky workaround. They key is to pick the black color inside the pic using color picker. I previously choose the color using color wheel. This is very useful. Thanks!

**Comment by Acly** [comment]
> They key is to pick the black color inside the pic using color picker.

Hm, I used color wheel though. I will have to recheck if it was 100% black or not.

> When I want to correct the hand, I select the whole figure, then erase all the generated area except the hand.

I'm a bit surprised at this example as an argument for more context. What you describe has been an exercise in futility in my experience. SD is very bad at hands, context or not. You can get lucky, but you may not, and I prefer more reliable methods. Rough painting by hand, or using a photo or line art (via CN) reference, and low-strength inpainting to make it fit in. Having more resolution available can also help, the problem that it looks a bit too detailed will be solved by subsequent upscale, or just a large area img2img at very low strength.

Global context is important for composition, colors and anatomy on a broad, coarse level, but has little to no impact on fine detail. Refining at low strength does not benefit, as the broad/coarse information is already available from existing content. At higher strength information from different parts of the image becomes more important. The plugin generally uses a much larger area for 100% to accomodate this. But for <100% it is a fixed (configurable) value. It might be an idea to scale padding with strength?

**Comment by nakanakapoi** [comment]
While the selection mask is selected, my color wheel become grayed out. I probably did something wrong.

Yes, SD is bad at hand. Other than context, the resolution must be optimum. With big enough context and correct resolution, I always get good hand at least one out of eight generation.

**Comment by Miraihi** [comment]
@nakanakapoi 
> I get it that previous approach by similar plugin (auto-sd-paint-ext) is cumbersome.
Just want to point out that there is [a competent rewrite of auto-sd-paint-ext](https://github.com/DrCyanide/cyanic-sd-krita) if you need this exact functionality (And some more).

**Comment by nakanakapoi** [comment]
> To be explicit, what denoising strength do you use? What is the optimum resolution?
> 
> The goal should be that you don't have to be aware of what resolution you're at, don't have to manually indicate it, and get the optimum anyway. Ideally.

Yes. The direction that we should be heading is to not be aware of resolution. If this plugin can resolve this, it will be fantastic.

I've spent hours everday with SD for the past months making characters in head to toe view. In my case, the optimum resolution for the highest quality is 704x1408 for SDXL. Anything more than that the anatomy got distorted, especially if the character is muscular. I also train my own LoRa for body build specific for this resolution ratio.

For face detailing, the optimum area for inpainting is around 768x768 when using face lora that I train. Anything more, or sometime less than that will reduce the likeness. For this reason, I always upscale the whole pic before detailing the face.

Denoising strength always depend on circumstances. High denoise for more variation. Low denoise for less variation.

This is just my own experience. This will vary when using different model and LoRa.


> Just want to point out that there is [a competent rewrite of auto-sd-paint-ext](https://github.com/DrCyanide/cyanic-sd-krita) if you need this exact functionality (And some more) ASAP. It's in active development but already very usable, and much less prone to breaking by design (It uses the conventional API calls rather than internal SD values). If you're a power user the developer will appreciate your feedback.

Thank you for the introduction. However, my whole workflow depends on comfy. But I will definitely try it.

**Comment by buzzz3d** [comment]
> * Make a selection of target area
> * Krita automatically creates a "Selection Mask" layer, activate it

I must be doing something wrong, because when I make a selection with the rectangular selection tool, nothing changes in the Layers palette, I don't see a Selection Mask Layer appearing. Sorry, not familiar with Krita so there must be some basic thing I'm doing wrong here.

**Comment by nakanakapoi** [comment]
> I must be doing something wrong, because when I make a selection with the rectangular selection tool, nothing changes in the Layers palette, I don't see a Selection Mask Layer appearing. Sorry, not familiar with Krita so there must be some basic thing I'm doing wrong here.

You must enable it first. 

Select > show global selection mask

**Comment by buzzz3d** [comment]
Excellent, thanks

Op za 2 dec 2023 16:19 schreef nakanakapoi ***@***.***>:

**Comment by AvidSpline** [comment]
I just noticed the PR for this. Awesome. This will be super useful for things like in-context object removal, small detail work, or locking areas of the image while regenerating others, which I was having trouble getting to work quite right with the existing tools in some cases.

Any usage tips so we can try this out ?

It seems to trigger when you have the mask layer selected, but i see the sizes are always the same and it doesn't seem to behave differently, e.g.:

2023-12-03 09:21:45,932 DEBUG Using experimental selection mask Selection Mask, mask bounds: Bounds(x=156, y=209, width=202, height=66), image bounds: Bounds(x=156, y=209, width=202, height=66)

Is there a trick I'm missing or am I jumping the gun and this isn't done yet ?

<img width="466" alt="Screenshot 2023-12-03 094056 - Copy" src="https://github.com/Acly/krita-ai-diffusion/assets/152655489/9fea7c84-addd-4625-a18c-bbc8116d478f">

**Comment by Miraihi** [comment]
> There are 3 ways to generate with masks now, they build on top of each other:

YES! This is exactly how I wanted it to work. Just tried that, this is perfect.

**Comment by DanielAdolfsson** [comment]
It is a bit confusing. It changes how generating a selection behaves, so it's not very intuitive in my opinion.

What if we had a "current context" mask somewhere else in the UI (like a small thumbnail somewhere), and a button to "make current selection the active context". It would become a mask that is *not* stored as a layer.

**Comment by Acly** [comment]
Using the mask image as inpaint mask and selection as context was mostly motivated by the fact that the mask needs be more precise and usually has soft edges which are not well represented by selections. Meanwhile the context is just a box.

But one alternative I might try (if I find the time...) is to have a "Context Control Layer", utilizing the same UI as other control layers. Here you would simply put any kind of layer to indicate the context area (eg by filling a box). This would leave the way to select the mask area unchanged, via selection, just like with automatic context. And you can keep some layers for your manual context areas and reuse them when needed.

Adding it to the control layer menu would make it discoverable with minimal additional UI clutter.

**Comment by teadrinker** [comment]
Thanks a lot for this write-up!
this feature was not discoverable for me (I think if "Select/show global selection mask" was on by default it maybe would have)

However, I found it's not working that well if I want to remove unwanted details (or if you want to completely regenerate a part of an image), because for high strength values it don't care about the context, and for low/mid strength values it also interprets the unwanted detail as context/blueprint. So it's more like selective reinterpretation than inpainting. (clearing/removing the parts of the image don't help, as it is just interpreted as black)

Am I doing this wrong, or is there some other workaround?

I suspect this is just the behaviour of SD, I know in a1111 I got better results with the specific inpainting models, however result was still worse than dalle-2 inpainting.

**Comment by Acly** [comment]
This is now (finally) more official and discoverable as part of [Custom Generation](https://github.com/Acly/krita-ai-diffusion/wiki/Inpainting#context-area). I hope the new workflow is also slightly less tedious.

Note that the old method described here is obsolete, it will no longer work in [v1.14.0](https://github.com/Acly/krita-ai-diffusion/releases/tag/v1.14.0)

**Comment by nakanakapoi** [comment]
Just spend hours trying this new method. It work greats and save times. BTW they hacky workaround using the small brush still work in v1.14.0
